### PR TITLE
Add “—endless-trace-buffer” to switch to an endless buffer from a ring buffer while tracing.

### PR DESCRIPTION
**Pull request by chinmaygarde**

### CODE REVIEW

**Comment by eseidelGoogle** [comment]
@johnmccutchan @danrubel

File: runtime/dart_init.cc [hunk 1]
```diff
@@ -580,8 +585,18 @@ void InitDartVM() {
...
     PushBackAll(&args, kDartStartPausedArgs, arraysize(kDartStartPausedArgs));
 
-  if (settings.trace_startup)
+  if (settings.trace_startup) {
     PushBackAll(&args, kDartTraceStartupArgs, arraysize(kDartTraceStartupArgs));
+    // If we are tracing startup, make sure the trace buffer is endless so we
+    // don't lose early traces.
+    PushBackAll(&args, kDartEndlessTraceBufferArgs,
+                arraysize(kDartEndlessTraceBufferArgs));
+  }
```
Comment (johnmccutchan) [inline runtime/dart_init.cc]: else if (settings.endless_trace_buffer) ?

**Comment by johnmccutchan** [comment]
LGTM

**Comment by chinmaygarde** [comment]
Rewrote that logic a bit differently so we are not adding the same flags from two different spots in code.

**Comment by rmacnak-google** [comment]
Note there is also a third recorder, "startup," which uses a fixed buffer and drops events after it gets full. This may be useful if there are so many events that the endlessly growing buffer gets the app killed for using too much memory. Early on, it was the only way I could get a timeline out of my iPod. But the engine makes far fewer API calls now, API calls aren't recorded by default anymore, and the VM uses less memory to send the timeline, so the endless buffer is probably fine for the foreseeable future.

**Pull request merged**